File: cpp/Hash_Table/0018_4Sum.cpp
#include <vector>
#include <unordered_map>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
  /*
   * 1st solution two pointers
   * */
  vector <vector<int>> fourSum(vector<int> &nums, int target) {
    vector <vector<int>> res;
    sort(nums.begin(), nums.end());

    for (int i = 0; i < nums.size(); ++i) {
      if (i > 0 && nums[i] == nums[i - 1]) continue;

      for (int j = i + 1; j < nums.size(); ++j) {
        if (j > i + 1 && nums[j] == nums[j - 1]) continue;
        int left = j + 1, right = nums.size() - 1;

        while (right > left) {
          if (nums[i] + nums[j] > target - nums[left] - nums[right]) right--;
          else if (nums[i] + nums[j] < target - nums[left] - nums[right]) left++;
          else {
            res.emplace_back(vector < int > {nums[i], nums[j], nums[left], nums[right]});
            while (right > left && nums[right] == nums[right - 1]) right--;
            while (right > left && nums[left] == nums[left + 1]) left++;
            right--;
            left++;
          }
        }
      }
    }

    return res;
  }
};
//leetcode submit region end(Prohibit modification and deletion)

File: python3/0018-4sum.py
class Solution:
    def fourSum(self, nums: [int], target: int) -> [[int]]:
        res = []
        nums.sort()
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            for j in range(i + 1, len(nums)):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                left, right = j + 1, len(nums) - 1
                while left < right:
                    total = nums[i] + nums[j] + nums[left] + nums[right]
                    if total > target:
                        right -= 1
                    elif total < target:
                        left += 1
                    else:
                        res.append([nums[i], nums[j], nums[left], nums[right]])
                        while left != right and nums[left] == nums[left + 1]:
                            left += 1
                        while left != right and nums[right] == nums[right - 1]:
                            right -= 1
                        left += 1
                        right -= 1
        return res


nums = [1, 0, -1, 0, -2, 2]
target = 0
print(Solution().fourSum(nums, target))  # [[-2,-1,1,2],[-2,0,0,2],[-1,0,0,1]]

nums = [2, 2, 2, 2, 2]
target = 8
print(Solution().fourSum(nums, target))  # [[2,2,2,2]]
